google-cloud-bigtable/lib/google/cloud/bigtable/table_data_operations.rb

```diff
@@ -0,0 +1,483 @@
+# frozen_string_literal: true
+
+require "google/gax"
+require "google/cloud/bigtable/v2"
+require "google/cloud/bigtable/chunk_reader"
+require "google/cloud/bigtable/flat_row"
+require "google/cloud/bigtable/mutation_entry"
+require "google/cloud/bigtable/rows_reader"
+
+module Google
+  module Cloud
+    module Bigtable
+      GRPC_RETRYABLE_ERRORS = [
+        GRPC::DeadlineExceeded,
+        GRPC::Aborted,
+        GRPC::Unavailable,
+        GRPC::Core::CallError
```

Comment:
Why are you handling `GRPC::Core::CallError`? Is this needed?

Reply:
This is for smart read rows retry. On occurrence of above errors start reading from last read row key.

Reply:
My understanding was that `GRPC::Core::CallError` is an internal error and not something that should be handled by client code. Is this an error that should be handled in more places?

Reply:
Currenly using `GRPC::Core::CallError` for only read rows becuase
`GRPC::Core::CallError` error  is frequently occurring in `read_rows` with large `rows_limit` or large row range(more than 100K rows). Also `CallError` has empty cause code.

Here is stackstrace:
```
#<GRPC::Core::CallError: call#run_batch failed somehow>
/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:253:in `run_batch'
/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:253:in `remote_read'
/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:326:in `block in each_remote_read_then_finish'
/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:325:in `loop'
/grpc-1.10.0-x86_64-linux/src/ruby/lib/grpc/generic/active_call.rb:325:in `each_remote_read_then_finish'
/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/rows_reader.rb:35:in `each'
/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/rows_reader.rb:35:in `read'
/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/table_data_operations.rb:107:in `read_rows'
```
Origin of error is in C extention:

`#<GRPC::Core::CallError: call#run_batch failed somehow>` 
https://github.com/grpc/grpc/blob/88cb996ba37a7076c10cbb9a6e568bbcfd97ad7b/src/ruby/ext/grpc/rb_call.c#L866

Reply:
/cc @apolcyn

Reply:
@jiren How frequent is the CallError? Can you reliably reproduce?

Reply:
Yes, It is reproducible in every read rows call with large row limit.

It is very frequent for more then 100k rows read on Google VM and on a dev machine more then 10k rows read.

You can reproduce using data api pull request code or  https://github.com/jiren/google-cloud-ruby/tree/bigtable-data-api-wrapper and remove `GRPC::Core::CallError` from `google-cloud-bigtable/lib/google/cloud/bigtable/table_data_operations.rb` line: 17.

```ruby
require "google/cloud"
require "google/cloud/bigtable"

project_id = "project_id"
instance_id = "instance_id"
table_id = "scantest"
keyfile = "keyfile.json"

client = Google::Cloud.new(project_id, keyfile).bigtable(instance_id: instance-id)
#client = Google::Cloud.new(project_id).bigtable(instance_id: instance-id)
table = client.table("scantest")

count = 0
table.read_rows(rows_limit: 100000) do |row|
  count += 1
  p count if count % 1000 == 0
end
```

Reply:
In general, Cloud Bigtable needs to have retries handled by the client.  There are various ways that networking can fail in transient ways, and the client needs to gracefully handle those cases.

Reply:
GRPC 1.11.0 was just released, with a fix for CallError. Can you verify if you can remove GRPC::Core::CallError from your list of handled error using this release?

Reply:
Still getting same errors after updating GRPC to `1.11.0`

```
/.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/grpc-1.11.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:253:in `run_batch': call#run_batch failed somehow (GRPC::Core::CallError)
from /.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/grpc-1.11.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:253:in `remote_read'
from /.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/grpc-1.11.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:326:in `block in each_remote_read_then_finish'
from /.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/grpc-1.11.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:325:in `loop'
from /.rbenv/versions/2.4.0/lib/ruby/gems/2.4.0/gems/grpc-1.11.0-universal-darwin/src/ruby/lib/grpc/generic/active_call.rb:325:in `each_remote_read_then_finish'
from /work/big_table/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/rows_reader.rb:80:in `each'
from /work/big_table/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/rows_reader.rb:80:in `read'
from /work/big_table/google-cloud-ruby/google-cloud-bigtable/lib/google/cloud/bigtable/table_data_operations.rb:234:in `read_rows'
from read_test.rb:40:in `<main>'
```